File: android/app/src/main/java/studio/room211/diktat/Groq.kt
package studio.room211.diktat

import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/** Groq GPT-OSS obrada već transkribovanog teksta. */
object Groq {

    private const val CHAT_ENDPOINT =
        "https://api.groq.com/openai/v1/chat/completions"
    const val DEFAULT_TEXT_MODEL = "openai/gpt-oss-120b"

    class GroqException(message: String) : Exception(message)

    private fun response(conn: HttpURLConnection): String {
        val code = conn.responseCode
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        val text = stream?.bufferedReader()?.readText().orEmpty()
        if (code !in 200..299) throw GroqException(explain(code, text))
        return text
    }

    /** Provera autentifikacije bez slanja audio-snimka ili teksta. */
    fun testKey(cfg: Config): String {
        if (cfg.groqApiKey.isBlank()) throw GroqException("Groq API ključ nije podešen.")
        val conn = (URL("https://api.groq.com/openai/v1/models").openConnection()
            as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 15_000
            readTimeout = 30_000
            setRequestProperty("Authorization", "Bearer ${cfg.groqApiKey}")
            setRequestProperty("User-Agent", "Diktat/1.0")
        }
        return try {
            val reply = response(conn)
            val count = runCatching { JSONObject(reply).optJSONArray("data")?.length() ?: 0 }
                .getOrDefault(0)
            "Groq ključ radi ($count modela dostupno)"
        } finally {
            conn.disconnect()
        }
    }

    /** Obradi tekst bez slanja audio-snimka. */
    fun manipulateText(text: String, instruction: String, cfg: Config): String {
        val payload = JSONObject().apply {
            put("model", DEFAULT_TEXT_MODEL)
            put("messages", JSONArray()
                .put(JSONObject().put("role", "system")
                    .put("content", "Poštuj uputstvo i vrati samo konačan obrađen tekst."))
                .put(JSONObject().put("role", "user")
                    .put("content", "$instruction\n\nSirov transkript:\n$text")))
            put("temperature", 0.0)
            put("max_completion_tokens", 2048)
            put("top_p", 1)
            put("reasoning_effort", "medium")
            put("stream", false)
        }.toString().toByteArray(Charsets.UTF_8)
        val conn = (URL(CHAT_ENDPOINT).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            doOutput = true
            connectTimeout = 15_000
            readTimeout = 60_000
            setRequestProperty("Authorization", "Bearer ${cfg.groqApiKey}")
            setRequestProperty("Content-Type", "application/json")
            setRequestProperty("User-Agent", "Diktat/1.0")
            setFixedLengthStreamingMode(payload.size)
        }
        try {
            conn.outputStream.use { it.write(payload) }
            val reply = response(conn)
            cfg.addTraffic(
                payload.size.toLong(), reply.toByteArray().size.toLong(), 0.0, false,
            )
            val content = JSONObject(reply).getJSONArray("choices")
                .getJSONObject(0).getJSONObject("message").optString("content")
            return content.trim().ifBlank {
                throw GroqException("GPT-OSS je vratio prazan tekst za obradu.")
            }
        } finally { conn.disconnect() }
    }

    private fun explain(code: Int, detail: String) = when (code) {
        401, 403 -> "Groq je odbio API ključ (HTTP $code)."
        404 -> "Groq model ili endpoint ne postoji (HTTP 404)."
        429 -> "Groq je ograničio broj zahteva (HTTP 429)."
        else -> "Groq je vratio HTTP $code." + detail.take(180)
    }
}

File: dictate/groq.py
"""Groq GPT-OSS kao model za AI obradu vec prepoznatog teksta.

Ovde se ne salje zvuk. Drugo misljenje o snimku (Whisper pa spajanje sa
Google prepisom) je uklonjeno 22.09.2026; kako je radilo i zasto je
uklonjeno stoji u docs/provera-snimka.md. API kljuc se cita samo iz lokalnog
config-a.
"""

import json
import urllib.error
import urllib.request


CHAT_ENDPOINT = "https://api.groq.com/openai/v1/chat/completions"
DEFAULT_TEXT_MODEL = "openai/gpt-oss-120b"
USER_AGENT = "Diktat/1.0"


class GroqError(Exception):
    """Greška Groq poziva; obrada teksta ne sme da obori diktat."""


def _request_json(request, timeout: float):
    try:
        raw = urllib.request.urlopen(request, timeout=timeout).read()
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", "replace")[:240]
        raise GroqError(_http_message(exc.code, detail)) from exc
    except urllib.error.URLError as exc:
        raise GroqError(f"Nema veze sa Groq-om ({exc.reason}).") from exc
    except TimeoutError as exc:
        raise GroqError("Groq nije odgovorio na vreme.") from exc
    try:
        return json.loads(raw.decode("utf-8", "replace"))
    except json.JSONDecodeError as exc:
        raise GroqError("Groq je vratio neispravan odgovor.") from exc


def manipulate_text(text: str, cfg, instruction: str, timeout=60) -> str:
    """Obradi već prepoznat tekst bez slanja audio-snimka."""
    key = (cfg.get("groq_api_key") or "").strip()
    if not key:
        raise GroqError("Nema Groq API ključa za obradu teksta.")
    payload = {
        "model": DEFAULT_TEXT_MODEL,
        "messages": [
            {
                "role": "system",
                "content": (
                    "Obrađuješ diktirani tekst. Poštuj uputstvo doslovno i "
                    "vrati samo konačan tekst."
                ),
            },
            {
                "role": "user",
                "content": f"{instruction}\n\nSirov transkript:\n{text}",
            },
        ],
        "temperature": 0.0,
        "max_completion_tokens": int(cfg.get("groq_max_completion_tokens", 2048)),
        "top_p": 1,
        "reasoning_effort": cfg.get("groq_reasoning_effort", "medium"),
        "stream": False,
    }
    request = urllib.request.Request(
        CHAT_ENDPOINT,
        data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        headers={
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
            "User-Agent": USER_AGENT,
        },
    )
    response = _request_json(request, timeout)
    try:
        content = response["choices"][0]["message"]["content"]
        if isinstance(content, list):
            content = "".join(
                part.get("text", "") for part in content if isinstance(part, dict)
            )
        result = str(content or "").strip()
    except (KeyError, IndexError, TypeError) as exc:
        raise GroqError("Groq GPT-OSS nije vratio tekst za obradu.") from exc
    if not result:
        raise GroqError("Groq GPT-OSS je vratio prazan tekst za obradu.")
    return result


def _http_message(code: int, detail: str = "") -> str:
    if code in (401, 403):
        return "Groq je odbio API ključ (HTTP %d)." % code
    if code == 404:
        return "Groq model ili endpoint ne postoji (HTTP 404)."
    if code == 429:
        return "Groq je ograničio broj zahteva (HTTP 429)."
    return f"Groq je vratio HTTP {code}." + (f" {detail}" if detail else "")
